diff --git a/configure b/configure
index 04e489d..1d385fe 100755
--- a/configure
+++ b/configure
@@ -3244,18 +3244,18 @@ if [ -z "${CFG_HOST_ARCH}" ]; then
     	fi
     	CFG_HOST_ARCH=s390
     	;;
+    *:*:aarch64*|*:*:arm64*)
+        if [ "$OPT_VERBOSE" = "yes" ]; then
+            echo "    AArch64 (aarch64)"
+        fi
+        CFG_HOST_ARCH=aarch64
+        ;;
     *:*:arm*)
         if [ "$OPT_VERBOSE" = "yes" ]; then
             echo "    ARM (arm)"
         fi
         CFG_HOST_ARCH=arm
 	;;
-    *:*:aarch64*)
-        if [ "$OPT_VERBOSE" = "yes" ]; then
-            echo "    AArch64 (aarch64)"
-        fi
-        CFG_HOST_ARCH=aarch64
-        ;;
     Linux:*:sparc*)
         if [ "$OPT_VERBOSE" = "yes" ]; then
             echo "    Linux on SPARC"
