diff -up qt-everywhere-opensource-src-4.6.2/src/3rdparty/harfbuzz/src/harfbuzz-indic.cpp.me qt-everywhere-opensource-src-4.6.2/src/3rdparty/harfbuzz/src/harfbuzz-indic.cpp
--- qt-everywhere-opensource-src-4.6.2/src/3rdparty/harfbuzz/src/harfbuzz-indic.cpp.me	2011-07-11 15:59:29.000000000 +0200
+++ qt-everywhere-opensource-src-4.6.2/src/3rdparty/harfbuzz/src/harfbuzz-indic.cpp	2011-07-11 16:03:49.000000000 +0200
@@ -833,7 +833,7 @@ static const unsigned char indicPosition
     None, None, None, None,
     None, None, None, Post,
 
-    Post, None, Below, None,
+    Pre, None, Below, None,
     None, Post, None, None,
     None, None, None, None,
     None, None, Post, Post,
